src/login/pam_systemd.c

```diff
@@ -199,35 +199,56 @@ static bool display_is_local(const char *display) {
                 display[1] <= '9';
 }
 
-static int socket_from_display(const char *display, char **path) {
+static int socket_from_display(const char *display) {
+        _cleanup_free_ char *f = NULL;
         size_t k;
-        char *f, *c;
+        char *c;
+        union sockaddr_union sa;
+        socklen_t sa_len;
+        _cleanup_close_ int fd = -1;
+        int r;
 
         assert(display);
-        assert(path);
 
         if (!display_is_local(display))
                 return -EINVAL;
 
         k = strspn(display+1, "0123456789");
 
-        f = new(char, STRLEN("/tmp/.X11-unix/X") + k + 1);
+        /* Try abstract socket first. */
+        f = new(char, STRLEN("@/tmp/.X11-unix/X") + k + 1);
         if (!f)
                 return -ENOMEM;
 
-        c = stpcpy(f, "/tmp/.X11-unix/X");
+        c = stpcpy(f, "@/tmp/.X11-unix/X");
         memcpy(c, display+1, k);
         c[k] = 0;
 
-        *path = f;
+        r = sockaddr_un_set_path(&sa.un, f);
+        if (r < 0)
+                return r;
+        sa_len = r;
 
-        return 0;
+        fd = RET_NERRNO(socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC, 0));
+        if (fd < 0)
+                return fd;
+
+        r = RET_NERRNO(connect(fd, &sa.sa, sa_len));
+        if (r >= 0)
+                return TAKE_FD(fd);
+
+        if (r == -ENOENT) {
+                /* Try also non-abstract socket. */
+                memmove(sa.un.sun_path, sa.un.sun_path + 1, sa_len);
```

Comment:
Is `sa_len` correct here? I think it should be `strlen(sa.sun_path) - 1`.

Reply:
Right, except `strlen(sa.un.sun_path + 1)`, since we don't want the first NUL byte.

Reply:
But you want to copy the NUL byte at the end of the string forward one position. I think `strlen(sa.un.sun_path)` would be correct for that.

Reply:
Or equivalently, `strlen(sa.un.sun_path + 1) + 1`.

Reply:
Maybe it would be nicer to just call `r = sockaddr_un_set_path(&sa.un, f + 1); sa_len = r;` instead of using memmove().

Reply:
Now using `sockaddr_un_set_path()`